### ISSUE TITLE
HTTPS batch message format is wrong

### ISSUE THREAD

**Issue by james-hu** [Azure/azure-iot-sdk-java#67]
- **OS and version used:** Windows 10

- **Java runtime used:** 8

- **SDK version used:** com.microsoft.azure.sdk.iot:iot-device-client:1.0.21

# Description of the issue:
When using HTTPS as the transport, sometimes the SDK combines multiple messages into one HTTPS request. However, the body format in the request is not a valid JSON format, therefore, 400 bad request error is returned from IoT Hub.

I think the problem is at this line:`jsonMsg.append("\"" + msg.getBodyAsString() + "\",");` in HttpsBatchMessage.java.

# Code sample exhibiting the issue:
Send multiple messages one by one

# Console log of the issue:
Request:
`POST https://............................/messages/events?api-version=2016-02-03 HTTP/1.1
User-Agent: com.microsoft.azure.sdk.iot.iot-device-client/1.0.21
authorization: SharedAccessSignature .................
..................
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2
Connection: keep-alive
Content-Length: 8013

[{"body":"{"HistorySamples":[{"ItemName":"sensor","Quality":"Good","Time":"2017-03-22T13:32:28.075+08:00","Value":80}..........`

Response:
`HTTP/1.1 400 Bad Request
Content-Length: 152
Content-Type: application/json; charset=utf-8
Server: Microsoft-HTTPAPI/2.0
iothub-errorcode: IotHubFormatError
Date: Wed, 22 Mar 2017 05:33:35 GMT

{"Message":"ErrorCode:IotHubFormatError;BadRequest","ExceptionMessage":"Tracking ID:429d18027c43471db3756a7c5d81fbd6-G:7-TimeStamp:03/22/2017 05:33:35"}`

**Comment by JMayrbaeurl** [comment]
You are right. Currently only single plain strings can be used for the body of the message. Objects and arrays in body will result in invalid json, since com.microsoft.azure.sdk.iot.device.transport.https.HttpsBatchMessage.msgToJson(HttpsSingleMessage) will always pre- and suffix the body with ".

**Comment by JMayrbaeurl** [comment]
Test case to validate the bug is: tests.integration.com.microsoft.azure.sdk.iot.device.transport.HttpsBatchMessageIT.batchMessageSetsFieldsCorrectlyForObjects()
```java
    @Test
    public void batchMessageSetsFieldsCorrectlyForObjects() throws
            SizeLimitExceededException
    {
        String msgBytes0 = "{\"HistorySamples\":[{\"ItemName\":\"sensor\",\"Quality\":\"Good\",\"Time\":\"2017-03-22T13:32:28.075+08:00\",\"Value\":80}]}";
        Message msg0 = new Message(msgBytes0);
        String messageid0 = msg0.getMessageId();
        msg0.setProperty("prop-0", "value-0");
        HttpsSingleMessage httpsMsg0 = HttpsSingleMessage.parseHttpsMessage(msg0);

        byte[] msgBytes1 = { 48, 49, 50 };
        Message msg1 = new Message(msgBytes1);
        String messageid1 = msg1.getMessageId();
        msg1.setProperty("prop-1", "value-1");
        HttpsSingleMessage httpsMsg1 = HttpsSingleMessage.parseHttpsMessage(msg1);
        HttpsBatchMessage batch = new HttpsBatchMessage();
        batch.addMessage(httpsMsg0);
        batch.addMessage(httpsMsg1);

        // JSON body with whitespace removed.
        String testBatchBody = new String(batch.getBody(), UTF8).replaceAll("\\s", "");

        String expectedBatchBody = "["
                + "{\"body\":" + "{\"HistorySamples\":[{\"ItemName\":\"sensor\",\"Quality\":\"Good\",\"Time\":\"2017-03-22T13:32:28.075+08:00\",\"Value\":80}]}"
                + "\"base64Encoded\":false,"
                + "\"properties\":{"
                +   "\"iothub-app-prop-0\":\"value-0\","
                +   "\"iothub-messageid\":\"" + messageid0 + "\"}},"
                + "{\"body\":\"012\","
                + "\"base64Encoded\":false,"
                + "\"properties\":{"
                +   "\"iothub-app-prop-1\":\"value-1\","
                +   "\"iothub-messageid\":\"" + messageid1 + "\"}}"
                + "]";


        assertThat(testBatchBody, is(expectedBatchBody));
    }
```

This can be fixed quite easily by checking the result of HttpsSingleMessage.getBodyAsString(). If the resulting string doesn't start with [ or { then it must be prefixed with " and suffixed with ".

**Comment by JMayrbaeurl** [comment]
And the patch in looks like
```java
    protected static String msgToJson(HttpsSingleMessage msg)
    {
        StringBuilder jsonMsg = new StringBuilder("{");
        // Codes_SRS_HTTPSBATCHMESSAGE_11_003: [The JSON object shall have the field "body" set to the raw message.]
        jsonMsg.append("\"body\":");
        String bodyString = msg.getBodyAsString();
        if (bodyString.startsWith("[") || bodyString.startsWith("{"))
        	jsonMsg.append(bodyString + ",");
        else
        	jsonMsg.append("\"" + msg.getBodyAsString() + "\",");
```

**Comment by james-hu** [comment]
Thanks for the quick reponse! 

Just for making sure, what is expected from the server? If the server expects "body" to be an object then the fix above should work, but if the server expects it to be a string, then we only need to escape the return value of msg.getBodyAsString(). If there can be an end-to-end integration test case that would be good.

**Comment by james-hu** [comment]
Is this issue being worked on?

**Comment by JMayrbaeurl** [comment]
Partner I'm working with is waiting for a fix! Any problems with the patch I've provided?

**Comment by timtay-microsoft** [comment]
Sorry for the delay on this fix. It is being worked on and will tentatively make the next release this Friday.

**Comment by anhashia** [comment]
Hi @JMayrbaeurl 

This fix for this is in **testing** phase. The patch that you have provided doesn't work because of limitation on Azure IoT Hub service side related to handling of JSON objects in HTTPS batching. So fix involves doing Base64 encoding of the data and setting the Base64 encoding flag to true. That should fix the issue. That is the same approach that is currently used in node.js device SDK which **doesn't** have this problem.
https://github.com/Azure/azure-iot-sdk-node/blob/master/device/transport/http/src/http.ts#L173

**Comment by prmathur-microsoft** [comment]
A fix for this issue has been checked in @566fbb7a503cfb42680354360add576230cd0358. Please give it try and maven packages should be available soon as well.

**Comment by prmathur-microsoft** [comment]
Maven package for the same is also released. Follow our release from https://github.com/Azure/azure-iot-sdk-java/releases/tag/2017-7-14
